features/steps/tabs.py

```diff
@@ -0,0 +1,79 @@
+# encoding: utf-8
+
+"""
+Step implementations for paragraph-related features
+"""
+
+from behave import given, then, when
+
+from docx import Document
+from docx.text.tabstop import TabStop
+from docx.shared import Inches
+
+from helpers import test_docx
+
+
+# given ===================================================
+@given('a tab_stops having {count} tab stops')
+def given_a_tab_stops_having_count_tab_stops(context, count):
+    count_int = {
+        '0': 0,
+        '3': 1
+    }[count]
+    document = Document(test_docx('tab-stops'))
+    paragraph_format = document.paragraphs[count_int].paragraph_format
+    context.tab_stops = paragraph_format.tab_stops
+
+
+# when ====================================================
+
+@when('I add a tab_stop')
+def when_i_add_a_tab_stop(context):
+    tab_stops = context.tab_stops
+    tab = tab_stops.add_tab()
+    tab.position = Inches(1.75)
+
+
+@when('I delete a tab_stop')
+def when_i_delete_a_tab_stop(context):
+    tab_stops = context.tab_stops
+    tab = tab_stops[1]
+    tab.delete()
```

Comment:
Steve,

When I tried to implement __delitem__() on the TabStop object, then 
tried del tab_stops[2], which is a TabStop item, I got a Python error 
saying that the TabStops object (the parent) doesn't support item 
deletion.  When I assigned tab_stops[2] to a variable and del'ed that 
variable, the underlying list was not affected, though no error message 
was displayed.

So I'm sticking with your recommendations:

tab = tab_stops[index]

tab_stops.remove(tab)
- and -
tab.clear()

These methods both work.  As always, your mileage may vary.

David



On 04/20/2016 02:09 PM, Steve Canny wrote:
>
> In features/steps/tabs.py 
> <https://github.com/python-openxml/python-docx/pull/261#discussion_r60469364>:
>
> > +
> > +
> > +# when ====================================================
> > +
> > +@when('I add a tab_stop')
> > +def when_i_add_a_tab_stop(context):
> > +    tab_stops = context.tab_stops
> > +    tab = tab_stops.add_tab()
> > +    tab.position = Inches(1.75)
> > +
> > +
> > +@when('I delete a tab_stop')
> > +def when_i_delete_a_tab_stop(context):
> > +    tab_stops = context.tab_stops
> > +    tab = tab_stops[1]
> > +    tab.delete()
>
> I notice there's a .clear() method on TabStop in the MS API. I'll 
> accept a .clear() method on our TabStop object into the API if you 
> like. That's consistent with the TabStops.clear() method, although 
> they call that .clear_all() in the MS API. I suppose, on balance, it 
> makes sense to rename TabStops.clear() to TabStops.clear_all(). That 
> reduces risk of confusion with variable names that only differ by a 
> single character (tab_stops.clear() vs. tab_stop.clear()). Then both 
> these methods would be named the same as in the MS API, which is a 
> consistency we've generally maintained throughout the library.
>
> .remove() has an established Python meaning in the context of a list, 
> so I wouldn't be wanting to choose a different name for that on the 
> TabStops object, which is essentially a custom list.
>
> Related to this, I did a little more research, and it turns out that 
> the magic method that gets called on |del lst[i]| is |__delitem__()|, 
> not |__del__()|. That makes me interested to try to make that work. 
> The advantage of |del| is that it supports slices.
>
> So maybe we can go back to the |del| syntax now instead of 
> |.remove()|. If you have trouble with the implementation I'll jump in 
> to get it working.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub 
> <https://github.com/python-openxml/python-docx/pull/261/files/dbfce4c2185a8718c44be6ddd260be5c5e6856ff..98ba6b5433023e9d8d3d32330d089063cad8e1bd#r60469364>
>

-- 
David K. Woods, Ph.D.
Researcher, Lead Transana Developer
Wisconsin Center for Education Research
University of Wisconsin, Madison
http://www.transana.org